### PR TITLE
feat(topology): support for logviewer for pods based on container

### PR DESCRIPTION
**Pull request by invincibleJai**
**Tracks:** https://github.com/janus-idp/backstage-plugins/issues/395

**Description:** Show the logs for pods associated with any k8s workload

**Screenshots/Gif**: 

image

image

Light Theme:

image

**TODO:**

- [x]  Should show `view logs`  in the resources tab for workloads in the pods section for sidepanel
- [x] On click on view logs, should open a modal with logs for the selected pod
- [x]  In the log component show a filter to search and should show a dropdown to select any container for pod
- [x] UI refinement (handle error and edge cases scenarios)
- [x]  Add Unit tests

image

### CODE REVIEW

File: plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.test.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    (usePodLogs as jest.Mock).mockReturnValue({
+      loading: false,
+      value: 'log data...',
+    });
+    const { queryByTestId } = render(<PodLogs podScope={podScopeData} />);
+    expect(queryByTestId('pod-log-banner')).not.toBeInTheDocument();
+    expect(queryByTestId('log-viewer')).toBeInTheDocument();
+  });
+
+  it('should render logviewer skelton', () => {
```
Comment (divyanshiGupta) [inline plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.test.tsx:37]: nit:
```suggestion
  it('should render logviewer skeleton', () => {
```

File: plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import { usePodLogs } from '../../../../hooks/usePodLogs';
+import { ContainerScope } from './types';
+
+type PodLogsProps = {
+  podScope: ContainerScope;
+};
+
+export const PodLogs = ({ podScope }: PodLogsProps) => {
+  const [loadingData, setLoadingData] = React.useState(true);
```
Comment (divyanshiGupta) [inline plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.tsx:16]:
```suggestion
  const [loadingData, setLoadingData] = React.useState<boolean>(true);
```

File: plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  const { value, error, loading } = usePodLogs({
+    podScope: podScope,
+  });
+
+  React.useEffect(() => {
+    if (!loading && loadingData) {
+      setLoadingData(false);
+    }
+  }, [loading, loadingData]);
```
Comment (divyanshiGupta) [inline plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.tsx:25]: Why can we not use `loading` itself instead of updating another variable based on its value and then using it?

File: plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogsDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+);
+
+type PodLogsDialogProps = {
+  podData: V1Pod;
+};
+
+export const PodLogsDialog = ({ podData }: PodLogsDialogProps) => {
+  const { clusters, selectedCluster } = React.useContext(K8sResourcesContext);
+  const classes = useStyles();
+  const [open, setOpen] = useState(false);
```
Comment (divyanshiGupta) [inline plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogsDialog.tsx:48]:
```suggestion
  const [open, setOpen] = useState<boolean>(false);
```

File: plugins/topology/src/hooks/useK8sObjectsResponse.test.ts [hunk 5]
```diff
@@ -44,4 +44,61 @@ describe('useK8sObjectsResponse', () => {
     expect(result.current.clusters).toEqual(['minikube']);
     expect(result.current.selectedClusterErrors).toEqual([]);
   });
+
+  it('should return k8sResourcesContextData with empty clusters if it does not exists', () => {
```
Comment (divyanshiGupta) [inline plugins/topology/src/hooks/useK8sObjectsResponse.test.ts:48]: nit:
```suggestion
  it('should return k8sResourcesContextData with empty clusters if it does not exist', () => {
```

File: plugins/topology/src/hooks/usePodLogs.test.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import { usePodLogs } from './usePodLogs';
+
+jest.mock('@backstage/core-plugin-api', () => ({
+  ...jest.requireActual('@backstage/core-plugin-api'),
+  useApi: jest.fn(),
+}));
+
+describe('usePodLogs', () => {
+  it('should return loading as true and value undefined initially', () => {
```
Comment (divyanshiGupta) [inline plugins/topology/src/hooks/usePodLogs.test.ts:13]:
```suggestion
  it('should return loading as true and value as undefined initially', () => {
```

File: plugins/topology/src/hooks/usePodLogs.test.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      }),
+    );
+
+    waitForNextUpdate();
+
+    expect(result.current.loading).toEqual(true);
+    expect(result.current.value).toBeUndefined();
+  });
+
+  it('should return loading as false and log text', async () => {
```
Comment (divyanshiGupta) [inline plugins/topology/src/hooks/usePodLogs.test.ts:35]:
```suggestion
  it('should return loading as false and value as log text', async () => {
```

**Review by divyanshiGupta — CHANGES_REQUESTED** [review]

File: plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.tsx (hunk 3 above)
Comment (invincibleJai) [reply plugins/topology/src/components/Topology/TopologySideBar/PodLogs/PodLogs.tsx]: you are right it's clean to move this usePodsLogs and the caller can rely on loading but the local state is needed as am making calls in intervals (used `useAsyncRetry`) and for each call loading turns to true and then false

File: plugins/topology/src/hooks/usePodLogs.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    [getLogs],
+  );
+
+  useInterval(() => retry(), intervalMs);
+
+  React.useEffect(() => {
+    if (!loading && loadingData) {
+      setLoadingData(false);
+    }
+  }, [loading, loadingData]);
```
Comment (divyanshiGupta) [inline plugins/topology/src/hooks/usePodLogs.ts:39]: @invincibleJai its not advised to use the state variable and its state update function in same useEffect as it might lead to an infinite loop. I understand the value will only be updated when `loadingData` is true but still wanted to mention this here. LMK wdyt.

**Review by divyanshiGupta — CHANGES_REQUESTED** [review]

**Comment by debsmita1** [comment]
verified this locally
LGTM
